### PR TITLE
Integrate data-loader for file import

### PR DESCRIPTION
**Pull request by loppear**
WIP: need to expose adminOpts to  AdminBase contructor.

### CODE REVIEW

File: src/adminBase.js [hunk 1]
```diff
@@ -11,6 +11,9 @@ import pluralize from 'pluralize'
 import _ from 'underscore'
 import morph from 'morph'
 import Promise from 'bluebird'
+import multer from 'multer';
+
+var upload = multer({ dest: process.cwd()+'/uploads/'})
```
Comment (alexjreich) [inline src/adminBase.js]: Why include these in adminBase and not in the data-uploader module?

File: src/adminBase.js [hunk 2]
```diff
@@ -120,15 +130,33 @@ export default class AdminBase extends HasModels {
...
+  uploadType() {
+    return this.opts.uploadType || null
+  }
+  /**
+   * Allow upload of models by file type
+   * @return {string}
+   */
+  uploadOptions() {
+    return this.opts.uploadOptions || {}
+  }
```
Comment (alexjreich) [inline src/adminBase.js:158]: Same as above - just out of curiosity, what are the use cases for uploads outside of data-loader?

**Comment by alexjreich** [comment]
Ok, so you’re thinking admin-ui includes the data upload functionality baked in? I can imagine a situation where we would want it to be separate so we can disable file uploading on models.  Also, is there a way to toggle upload on a model by model basis?  Again, not sure we want that by default.

**Comment by loppear** [comment]
It is optional for admin-ui, only used if an adminModel specifies an uploadType (so if you don't install data-loader and don't specify an uploadType, no impact). I started with it in data-loader, but I think it makes more sense as an optional dep for admin-ui than vice versa, the routes etc that data-loader would be providing are all about admin-ui.

File: src/adminBase.js (hunk 2 above)
Comment (loppear) [inline src/adminBase.js:158]: None, these are just the optional options to configure it if you are using both.

**Comment by alexjreich** [comment]
Ok, but are we using that convention anywhere else?  So far, we’ve been using a gather to enable features like that:

```
app.get(‘admin-ui’).adminModel(…)
app.get(‘searcher’).searchable(…)
```

Why not do the same thing?

**Comment by alexjreich** [comment]
@loppear yeah, I agree on routes, etc - the integration into admin-ui list makes sense.  Though the assumption is that we need to provide admin-ui CRUD to enable data upload, and I’m not sure that is true all the time.  Take Tracts for example - we probably want to disable admin-ui CRUD for that (since those should never be edited directly) but still want data upload capability.

So the question is whether upload should be its own menu item with a form where you select the destination model, or integrated as you currently have it.  I can see advantages/disadvantages both ways….

**Comment by loppear** [comment]
Either way works-ish, but two reasons:
1. admin-ui can do things with its templates (stick the upload button on the list view, rather than just in nav) given that it knows about upload. We could add gathers for 'adminModelButtons' instead though.
2. I see admin-ui as the aggregator of functionality: as I said above, one of them needs to know about the other, but does it make more sense for data-loader to know all about admin-ui or vice versa? (ala admin-ui<->users, users could have a config option for protecting admin-ui's routes, but that makes more sense as it is today with admin-ui controlling that - similarly, specifying "options for uploading in the admin-ui" reads better to me with the other adminModel options.)

**Comment by loppear** [comment]
I don't think we'd often have a totally standalone "pick a model" import, but I can see having non-admin upload pages (that are similarly just integration between a given route, template, and config for importFileToModel).

How about if I move the multer setup over to data-loader.

File: src/adminBase.js (hunk 2 above)
Comment (alexjreich) [inline src/adminBase.js:158]: Just FYI - docs are a bit opaque here?  Is the return string supposed to be a model name?  What are the uploadOptions?

**Comment by loppear** [comment]
As far as removing admin-ui CRUD for Tract, I'd probably vote for keeping the admin view of the model in there (for sanity-checking, but add options to AdminModel for whether you want to allow edit/delete.

**Pull request merged**